### PR TITLE
feat: add `finder` option for choosing a preferred finder backend (#129)

### PR DESCRIPTION
**Pull request by idanarye**

### CODE REVIEW

File: doc/obsidian.txt [hunk 1]
```diff
@@ -180,6 +180,14 @@ Using `lazy.nvim` <https://github.com/folke/lazy.nvim>, for example:
...
         -- Optional, set to true to force ':ObsidianOpen' to bring the app to the foreground.
         open_app_foreground = false,
+
+	-- Optional, by default commands like `:ObsidianSearch` will attempt -
+	-- in that order - telescope.nvim, fzf-lua, and fzf.nvim, and use the
+	-- first one they find. By setting this option to your preferred
+	-- finder, you can attempt it first. Note that if the specified finder
+	-- is not installed, or if it the command does not support it, the
+	-- remaining finders will be attempted in the original order.
+	finder = "fzf-lua",
```
Comment (epwalsh) [inline doc/obsidian.txt:190]: This is actually updated automatically from the README
```suggestion
```

File: README.md [hunk 2]
```diff
@@ -149,6 +149,14 @@ return {
...
     -- Optional, set to true to force ':ObsidianOpen' to bring the app to the foreground.
     open_app_foreground = false,
+
+	-- Optional, by default commands like `:ObsidianSearch` will attempt -
+	-- in that order - telescope.nvim, fzf-lua, and fzf.nvim, and use the
+	-- first one they find. By setting this option to your preferred
+	-- finder, you can attempt it first. Note that if the specified finder
+	-- is not installed, or if it the command does not support it, the
+	-- remaining finders will be attempted in the original order.
+	finder = "fzf-lua",
```
Comment (epwalsh) [inline README.md:159]: The indentation looks off here
```suggestion

    -- Optional, by default commands like `:ObsidianSearch` will attempt -
    -- in that order - telescope.nvim, fzf-lua, and fzf.nvim, and use the
    -- first one they find. By setting this option to your preferred
    -- finder, you can attempt it first. Note that if the specified finder
    -- is not installed, or if it the command does not support it, the
    -- remaining finders will be attempted in the original order.
    finder = "fzf-lua",
```

File: lua/obsidian/command.lua [hunk 3]
```diff
@@ -287,70 +295,73 @@ command.insert_template = function(client, data)
...
+                apply_template(template[1])
+              end)
+              return true
+            end,
+          }
+          require("telescope.builtin").find_files(opts)
+        end
+        choose_template()
+        return true
+      end
```
Comment (epwalsh) [inline lua/obsidian/command.lua:319]: I think we should explicitly return false if `has_telescope` is false, and same for the others below. Just for clarity.

**Review by epwalsh — CHANGES_REQUESTED** [review]
Hey @idanarye, I really like the approach you took. Only minor comments. 👍

File: lua/obsidian/command.lua (hunk 3 above)
Comment (idanarye) [reply lua/obsidian/command.lua:319]: Should I also verify in `run_first_supported` that the returned value is boolean?

Comment (idanarye) [reply lua/obsidian/command.lua:306]: Another option is to throw a magic value from skipped implementations. That way `run_first_supported` can return the value of the implementation that succeeded.

Comment (epwalsh) [reply lua/obsidian/command.lua:306]: > Another option is to throw a magic value from skipped implementations. That way run_first_supported can return the value of the implementation that succeeded.

I like that idea 👍

**Review by epwalsh — COMMENTED** [review]
Thanks @idanarye, just a minor suggestion to make `util.run_first_supported` more readable.

File: lua/obsidian/util.lua [hunk 4]
```diff
@@ -433,4 +433,72 @@ util.working_day_before = function(time)
...
+  for _, impl_name in ipairs(order) do
+    local impl_function = implementations[impl_name]
+    if impl_function then
+      local result = { pcall(impl_function) }
+      if result[1] then
+        return select(2, unpack(result))
+      elseif result[2] == IMPLEMENTATION_UNAVAILABLE then
+        table.insert(unavailable, impl_name)
+      else
+        error(result[2])
```
[the remark this answers is not in the document]
Comment (idanarye) [reply lua/obsidian/util.lua:465]: But then we won't be able to use multiple returns from the implementation functions. `_run_with_finder_backend` doesn't use that (it doesn't even return a single value), but `run_first_supported` is a library function so I tried to make it as generic as possible.

Then again, maybe I'm just prematurely generalizing things...

[the remark this answers is not in the document]
Comment (epwalsh) [reply lua/obsidian/util.lua:465]: Oooh, I see. Ok this is fine with me

**Comment by epwalsh** [comment]
Looks like there's a styling and linting issue

**Comment by idanarye** [comment]
@epwalsh There is one warning left, but it was not introduced by this PR (`main` has it: https://github.com/epwalsh/obsidian.nvim/actions/runs/4939421696/jobs/8830150709)

**Comment by epwalsh** [comment]
Ah, sorry about that. Fixed.

**Review by epwalsh — APPROVED** [review]
LGTM!

**Pull request merged**